### ISSUE TITLE
Key Exchange Negotiation Failed

### ISSUE THREAD

**Issue by LoneWolfPlanet** [sshnet/SSH.NET#858]
Hello sirs.
I have verified  both supported Key Exchange Methods in the server and client using "ssh -Q kex" command 
and found that they the same methods .
But still I'm encountering the issue below.

**Code**
using(SftpClient client = new SftpClient( "hostname", "username" , "password")){
      client.Connect();
}

**Stack Trace** 
Key exchange negotiation failed.
   at Renci.SshNet.Security.KeyExchange.Finish() + 0x71
   at Renci.SshNet.Security.KeyExchangeECCurve25519.Finish() + 0xe
   at Renci.SshNet.Security.KeyExchangeECCurve25519.Session_KeyExchangeEcdhReplyMessageReceived(Object, MessageEventArgs`1) + 0x3f
   at Renci.SshNet.Session.OnKeyExchangeEcdhReplyMessageReceived(KeyExchangeEcdhReplyMessage) + 0x60
   at Renci.SshNet.Messages.Transport.KeyExchangeEcdhReplyMessage.Process(Session) + 0x14
   at Renci.SshNet.Session.MessageListener() + 0x53
   at Renci.SshNet.Session.WaitOnHandle(WaitHandle, TimeSpan) + 0x100
   at Renci.SshNet.Session.WaitOnHandle(WaitHandle) + 0x16
   at Renci.SshNet.Session.Connect() + 0x2d7
   at Renci.SshNet.BaseClient.CreateAndConnectSession() + 0x12c
   at Renci.SshNet.BaseClient.Connect() + 0x9b

**Comment by IgorMilavec** [comment]
Can you try to disable ECC25519 to see if it is able to verify using other methods?
```C#

ConnectionInfo connectionInfo = new ConnectionInfo(
    host,
    port,
    userName,
    authenticationMethods);

connectionInfo.KeyExchangeAlgorithms.Remove("curve25519-sha256");
connectionInfo.KeyExchangeAlgorithms.Remove("curve25519-sha256@libssh.org");

SftpClient client = new SftpClient(connectionInfo)
```

**Comment by ckorn** [comment]
I am getting this exception, too.
Your code did not solve it.

```
Key exchange negotiation failed.
   at Renci.SshNet.Session.WaitOnHandle(WaitHandle waitHandle, TimeSpan timeout)
   at Renci.SshNet.Session.WaitOnHandle(WaitHandle waitHandle)
   at Renci.SshNet.Session.Connect()
   at Renci.SshNet.BaseClient.CreateAndConnectSession()
   at Renci.SshNet.BaseClient.Connect()
   at Logic.AutoScpManagement.AutoScpWatcher.ConnectClient(ConnectionInfo connectionInfo, SftpClient client, CancellationToken cancellationToken) in C:\Users\chris\source\repos\AutoScpWatcher\Logic\AutoScpManagement\AutoScpWatcher.cs:line 170

```

**Comment by ckorn** [comment]
This is the sshd debug log:
```
Aug 04 20:08:46 latte sshd[15871]: debug1: Local version string SSH-2.0-OpenSSH_8.9p1 Ubuntu-3
Aug 04 20:08:46 latte sshd[15871]: debug1: Remote protocol version 2.0, remote software version Renci.SshNet.SshClient.0.0.1
Aug 04 20:08:46 latte sshd[15871]: debug1: compat_banner: no match: Renci.SshNet.SshClient.0.0.1
Aug 04 20:08:46 latte sshd[15871]: debug2: fd 4 setting O_NONBLOCK
Aug 04 20:08:46 latte sshd[15871]: debug3: ssh_sandbox_init: preparing seccomp filter sandbox
Aug 04 20:08:46 latte sshd[15871]: debug2: Network child is on pid 15872
Aug 04 20:08:46 latte sshd[15871]: debug3: preauth child monitor started
Aug 04 20:08:46 latte sshd[15871]: debug3: privsep user:group 106:65534 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: permanently_set_uid: 106/65534 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: ssh_sandbox_child: setting PR_SET_NO_NEW_PRIVS [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: ssh_sandbox_child: attaching seccomp filter program [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: list_hostkey_types: rsa-sha2-512,rsa-sha2-256,ssh-rsa,ecdsa-sha2-nistp256,ssh-ed25519 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: send packet: type 20 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: SSH2_MSG_KEXINIT sent [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: receive packet: type 20 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: SSH2_MSG_KEXINIT received [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: local server KEXINIT proposal [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: KEX algorithms: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,sntrup761x25519-sha512@op>
Aug 04 20:08:46 latte sshd[15871]: debug2: host key algorithms: rsa-sha2-512,rsa-sha2-256,ssh-rsa,ecdsa-sha2-nistp256,ssh-ed25519 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: ciphers ctos: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: ciphers stoc: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: MACs ctos: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com>
Aug 04 20:08:46 latte sshd[15871]: debug2: MACs stoc: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com>
Aug 04 20:08:46 latte sshd[15871]: debug2: compression ctos: none,zlib@openssh.com [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: compression stoc: none,zlib@openssh.com [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: languages ctos:  [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: languages stoc:  [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: first_kex_follows 0  [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: reserved 0  [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: peer client KEXINIT proposal [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: KEX algorithms: ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group-exchange-sha1,>
Aug 04 20:08:46 latte sshd[15871]: debug2: host key algorithms: ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,ssh-rsa,ssh-dss [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: ciphers ctos: aes256-ctr,3des-cbc,aes128-cbc,aes192-cbc,aes256-cbc,blowfish-cbc,twofish-cbc,twofish192-cbc,twofish128-cbc,twofish256-cbc,arcfour,>
Aug 04 20:08:46 latte sshd[15871]: debug2: ciphers stoc: aes256-ctr,3des-cbc,aes128-cbc,aes192-cbc,aes256-cbc,blowfish-cbc,twofish-cbc,twofish192-cbc,twofish128-cbc,twofish256-cbc,arcfour,>
Aug 04 20:08:46 latte sshd[15871]: debug2: MACs ctos: hmac-md5,hmac-md5-96,hmac-sha1,hmac-sha1-96,hmac-sha2-256,hmac-sha2-256-96,hmac-sha2-512,hmac-sha2-512-96,hmac-ripemd160,hmac-ripemd16>
Aug 04 20:08:46 latte sshd[15871]: debug2: MACs stoc: hmac-md5,hmac-md5-96,hmac-sha1,hmac-sha1-96,hmac-sha2-256,hmac-sha2-256-96,hmac-sha2-512,hmac-sha2-512-96,hmac-ripemd160,hmac-ripemd16>
Aug 04 20:08:46 latte sshd[15871]: debug2: compression ctos: none [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: compression stoc: none [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: languages ctos:  [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: languages stoc:  [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: first_kex_follows 0  [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: reserved 0  [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: kex: algorithm: ecdh-sha2-nistp256 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: kex: host key algorithm: ssh-ed25519 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: kex: client->server cipher: aes256-ctr MAC: hmac-sha1 compression: none [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: kex: server->client cipher: aes256-ctr MAC: hmac-sha1 compression: none [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: expecting SSH2_MSG_KEX_ECDH_INIT [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: receive packet: type 30 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: SSH2_MSG_KEX_ECDH_INIT received [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: mm_sshkey_sign: entering [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: mm_request_send: entering, type 6 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: mm_sshkey_sign: waiting for MONITOR_ANS_SIGN [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: mm_request_receive_expect: entering, type 7 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: mm_request_receive: entering [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: mm_request_receive: entering
Aug 04 20:08:46 latte sshd[15871]: debug3: monitor_read: checking request 6
Aug 04 20:08:46 latte sshd[15871]: debug3: mm_answer_sign: entering
Aug 04 20:08:46 latte sshd[15871]: debug3: mm_answer_sign: ssh-ed25519 KEX signature len=83
Aug 04 20:08:46 latte sshd[15871]: debug3: mm_request_send: entering, type 7
Aug 04 20:08:46 latte sshd[15871]: debug2: monitor_read: 6 used once, disabling now
Aug 04 20:08:46 latte sshd[15871]: debug3: send packet: type 31 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug3: send packet: type 21 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug2: ssh_set_newkeys: mode 1 [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: rekey out after 4294967296 blocks [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: SSH2_MSG_NEWKEYS sent [preauth]
Aug 04 20:08:46 latte sshd[15871]: debug1: expecting SSH2_MSG_NEWKEYS [preauth]
Aug 04 20:08:46 latte sshd[15871]: Connection closed by 192.168.0.24 port 64288 [preauth]
```

**Comment by ckorn** [comment]
I deleted my comments because my issue was just a wrong host key fingerprint.

**Comment by WojciechNagorski** [comment]
Fixed by #1177 and #1180

**Comment by WojciechNagorski** [comment]
Version 2023.0.0 has been published https://www.nuget.org/packages/SSH.NET/2023.0.0
